crates/holochain/src/core/workflow/app_validation_workflow/validation_tests.rs

```diff
@@ -0,0 +1,419 @@
+        move |api: BoxApi, ()| {
+            let entry_def_id: EntryDefId = entry_def_b.id.clone();
+            let entry = Entry::app(().try_into().unwrap()).unwrap();
+            let hash = api.create(CreateInput::new(
+                entry_def_id,
+                entry,
+                ChainTopOrdering::default(),
+            ))?;
+            Ok(hash)
+        }
+    };
+
+    let (events_tx, mut events_rx) = tokio::sync::mpsc::channel(100);
+    const ZOME_A_0: &'static str = "ZOME_A_0";
+    const ZOME_A_1: &'static str = "ZOME_A_1";
+    const ZOME_B_0: &'static str = "ZOME_B_0";
+    const ZOME_B_1: &'static str = "ZOME_B_1";
+
+    const ALICE: &'static str = "ALICE";
+    const BOB: &'static str = "BOB";
+
+    #[derive(Debug, Hash, PartialEq, Eq, Clone)]
+    struct Event {
+        header: HeaderLocation,
+        op_type: DhtOpType,
+        called_zome: &'static str,
+        with_zome_index: Option<ZomeId>,
+        with_entry_def_index: Option<EntryDefIndex>,
+    }
+
+    impl Default for Event {
+        fn default() -> Self {
+            Self {
+                header: Default::default(),
+                op_type: DhtOpType::RegisterAgentActivity,
+                called_zome: Default::default(),
+                with_zome_index: Default::default(),
+                with_entry_def_index: Default::default(),
+            }
+        }
+    }
+
+    #[derive(Debug, Hash, PartialEq, Eq, Clone, Default)]
+    struct HeaderLocation {
+        agent: &'static str,
+        header_type: String,
+        seq: u32,
+    }
+
+    impl std::fmt::Display for Event {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            match (&self.with_zome_index, &self.with_entry_def_index) {
+                (Some(z), None) => write!(
+                    f,
+                    "{}:{}:{}:zome_id({})",
+                    self.called_zome, self.op_type, self.header, z
+                ),
+                (Some(z), Some(e)) => write!(
+                    f,
+                    "{}:{}:{}:zome_id({}):entry_id({})",
+                    self.called_zome, self.op_type, self.header, z, e.0
+                ),
+                _ => write!(f, "{}:{}:{}", self.called_zome, self.op_type, self.header),
+            }
+        }
+    }
+
+    impl std::fmt::Display for HeaderLocation {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            write!(f, "{}:{}:{}", self.agent, self.header_type, self.seq)
+        }
+    }
+
+    impl HeaderLocation {
+        fn new(header: impl Into<Header>, agents: &HashMap<AgentPubKey, &'static str>) -> Self {
+            let header = header.into();
+            Self {
+                agent: agents.get(header.author()).unwrap(),
+                header_type: header.header_type().to_string(),
+                seq: header.header_seq(),
+            }
+        }
+
+        fn expect(agent: &'static str, header_type: HeaderType, seq: u32) -> Self {
```

Comment:
```suggestion
        fn expected(agent: &'static str, header_type: HeaderType, seq: u32) -> Self {
```

just kinda weird to call something `expect()` since it's so common on Option and Result